clap_derive/src/derives/from_argmatches.rs

```diff
@@ -239,3 +249,171 @@ pub fn gen_constructor(
+                    Ty::Option => quote_spanned!( ty.span()=> Some(val) ),
+                    _ => quote_spanned!( ty.span()=> val ),
+                };
+
+                quote_spanned! { kind.span()=>
+                    if let Some(val) = {
+                        let (name, subcmd) = matches.subcommand();
+                        <#subcmd_type as ::clap::Subcommand>::from_subcommand(
+                            name,
+                            subcmd
+                        )
+                    } {
+                        self. #field_name = #val;
+                    }
+                }
+            }
+
+            Kind::Flatten => quote_spanned! { kind.span()=>
+                ::clap::FromArgMatches::update_from_arg_matches(&mut self. #field_name, matches);
+            },
+
+            Kind::Skip(_) => quote!(),
+
+            Kind::Arg(ty) => {
+                use self::ParserKind::*;
+
+                let parser = attrs.parser();
+                let func = &parser.func;
+                let span = parser.kind.span();
+                let (value_of, values_of, mut parse) = match *parser.kind {
+                    FromStr => (
+                        quote_spanned!(span=> value_of),
+                        quote_spanned!(span=> values_of),
+                        func.clone(),
+                    ),
+                    TryFromStr => (
+                        quote_spanned!(span=> value_of),
+                        quote_spanned!(span=> values_of),
+                        quote_spanned!(func.span()=> |s| #func(s).unwrap()),
+                    ),
+                    FromOsStr => (
+                        quote_spanned!(span=> value_of_os),
+                        quote_spanned!(span=> values_of_os),
+                        func.clone(),
+                    ),
+                    TryFromOsStr => (
+                        quote_spanned!(span=> value_of_os),
+                        quote_spanned!(span=> values_of_os),
+                        quote_spanned!(func.span()=> |s| #func(s).unwrap()),
+                    ),
+                    FromOccurrences => (
+                        quote_spanned!(span=> occurrences_of),
+                        quote!(),
+                        func.clone(),
+                    ),
+                    FromFlag => (quote!(), quote!(), func.clone()),
+                };
+
+                let flag = *attrs.parser().kind == ParserKind::FromFlag;
+                let occurrences = *attrs.parser().kind == ParserKind::FromOccurrences;
+                let name = attrs.cased_name();
+
+                let field_value = match **ty {
+                    Ty::Bool => quote_spanned! { ty.span()=>
+                        self. #field_name || matches.is_present(#name)
+                    },
+
+                    Ty::Option => {
+                        if attrs.is_enum() {
+                            if let Some(subty) = subty_if_name(&field.ty, "Option") {
```

Comment:
Ty::Option garantees the the wrapper type is `Option`. You can use `ty::sub_type` here.
```suggestion
                            if let Some(subty) = sub_type(&field.ty) {
```